ndcube/ndcube.py

```diff
@@ -4,74 +4,195 @@
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
+            The edges argument helps in returning ``pixel_edges``
+            instead of ``pixel_values``. Default value is False,
+            which returns ``pixel_values``. True return ``pixel_edges``
+
+        Returns
+        -------
+        axes_coords
+            `~astropy.units.Quantity` or iterable thereof for all requested
+            world axes, units determined by the wcs.  # TODO: Should we mandate namedtuple?
+
+        Example
+        -------
+        >>> NDCube.all_world_coords(('lat', 'lon')) # doctest: +SKIP
+        >>> NDCube.all_world_coords(2) # doctest: +SKIP
+
+        """
+
+    @abc.abstractmethod
+    def crop(self,
+             lower_corner: Iterable[Union[u.Quantity, float]],
+             upper_corner: Iterable[Union[u.Quantity, float]] = None,
+             units: Iterable[u.Unit] = None,
+             wcs: BaseHighLevelWCS = None):
         """
         Crops an NDCube given minimum values and interval widths along axes.
 
         Parameters
         ----------
-        lower_corner: iterable of `astropy.units.Quantity` or `float`
+        lower_corner
             The minimum desired values along each relevant axis after cropping
             described in physical units consistent with the NDCube's wcs object.
             The length of the iterable must equal the number of data dimensions
             and must have the same order as the data.
 
-        interval_widths: iterable of `astropy.units.Quantity` or `float`
-            The width of the region of interest in each dimension in physical
-            units consistent with the NDCube's wcs object. The length of the
-            iterable must equal the number of data dimensions and must have
-            the same order as the data. This argument will be removed in versions
-            2.0, please use upper_corner argument.
-
-        upper_corner: iterable of `astropy.units.Quantity` or `float`
+        upper_corner
             The maximum desired values along each relevant axis after cropping
             described in physical units consistent with the NDCube's wcs object.
             The length of the iterable must equal the number of data dimensions
             and must have the same order as the data.
 
-        units: iterable of `astropy.units.quantity.Quantity`, optionnal
+        units
             If the inputs are set without units, the user must set the units
             inside this argument as `str`.
             The length of the iterable must equal the number of data dimensions
             and must have the same order as the data.
 
+        wcs
+            The WCS to use to calculate the pixel coordinates based on the input.
+
         Returns
         -------
         result: NDCube
         """
 
 
```

Comment:
```suggestion
abc.abstractmethod
    def crop_by_values(self, lower_corner, upper_corner, units=None, wcs=None):
        """
        Crops an NDCube given lower and upper real world bounds for each real world axis.
        
        Parameters
        ----------
        lower_corner: iterable whose elements are None, `astropy.units.Quantity` or `float`
            An iterable of length-1 Quantities or floats, representing
            the real world coordinate values of the lower corner of
            the region of interest.
            These are input to `astropy.wcs.WCS.world_to_array_index_values`
            so their number and order must be compatible with the API of that method,
            i.e. they must be in world axis order.
            Alternatively, None, can be provided instead of a Quantity or float.
            In this case, the corresponding array axes will be cropped starting from
            0th array index.
        upper_corner: iterable whose elements are None, `astropy.units.Quantity` or `float`
            An iterable of length-1 Quantities or floats, representing
            the real world coordinate values of the upper corner of
            the region of interest.
            These are input to `astropy.wcs.WCS.world_to_array_index_values`
            so their number and order must be compatible with the API of that method,
            i.e. they must be in world axis order.
            Alternatively, None, can be provided instead of a Quantity or float.
            In this case, the corresponding array axes will be cropped to include
            the final array index.
        units: iterable of `astropy.units.Unit`
            The unit of the corresponding entries in lower_corner and upper_corner.
            Must therefore be the same length as lower_corner and upper_corner.
            Only used if the corresponding type is not a `astropy.units.Quantity`.
        wcs: `astropy.wcs.wcsapi.BaseLowLevelWCS`
            The WCS object to used to convert the world values to array indices.
            Although technically this can be any valid WCS, it will typically be
            self.wcs, self.extra_coords.wcs, or self.combined_wcs, combing both
            the WCS and extra coords.
            Default=self.wcs
        
        Returns
        -------
        result: `ndcube.NDCube`
        """
```

Reply:
The WCS object to used to convert  -> The WCS object _is_ used to convert

Reply:
I wonder where the `crop_by_values`  idea came from? I remember it being discussed before. Is it due to the fact you crop by read world values rather than data array indices? What about  `crop_by_wcs`  or something? or maybe that doesn't make sense.